### PR TITLE
Companion for #8615: enrich metadata with type information

### PR DESCRIPTION
**Pull request by ascjones**
Companion for https://github.com/paritytech/substrate/pull/8615.

Cumulus companion: https://github.com/paritytech/cumulus/pull/564

# todo:

- [x] a `scale-info` release (https://github.com/paritytech/scale-info/issues/89), 
- [x] merge https://github.com/paritytech/grandpa-bridge-gadget/pull/218

First look at json generated metadata: https://gist.github.com/ascjones/726caaf29eff76ca72f142f419e9c6a9

### CODE REVIEW

File: bridges/bin/millau/runtime/src/lib.rs [hunk 1]
```diff
@@ -460,7 +460,7 @@ impl_runtime_apis! {
 
 	impl sp_api::Metadata<Block> for Runtime {
 		fn metadata() -> OpaqueMetadata {
-			Runtime::metadata().into()
+			OpaqueMetadata::new(Runtime::metadata().into())
```
Comment (thiolliere) [inline bridges/bin/millau/runtime/src/lib.rs:463]: those change will need to be applied in https://github.com/paritytech/parity-bridges-common too
cc @HCastano

File: bridges/bin/runtime-common/Cargo.toml [hunk 2]
```diff
@@ -11,6 +11,7 @@ license = "GPL-3.0-or-later WITH Classpath-exception-2.0"
 codec = { package = "parity-scale-codec", version = "2.0.0", default-features = false, features = ["derive"] }
 ed25519-dalek = { version = "1.0", default-features = false, optional = true }
 hash-db = { version = "0.15.2", default-features = false }
+scale-info = { version = "0.10.0", default-features = false, features = ["derive"] }
```
Comment (thiolliere) [inline bridges/bin/runtime-common/Cargo.toml:14]: probably better to activate std features on std below.

File: runtime/parachains/src/lib.rs [hunk 3]
```diff
@@ -21,6 +21,8 @@
 //! of the other modules.
 
 #![cfg_attr(not(feature = "std"), no_std)]
+// `construct_runtime!` does a lot of recursion and requires us to increase the limit to 256.
+#![recursion_limit = "256"]
```
Comment (thiolliere) [inline runtime/parachains/src/lib.rs:25]: sidenote I thought construct_runtime didn't use that much recursion anymore actually.

File: runtime/polkadot/src/lib.rs [hunk 4]
```diff
@@ -833,7 +833,7 @@ where
 
 parameter_types! {
 	pub const ParathreadDeposit: Balance = 500 * DOLLARS;
-	pub const QueueSize: usize = 2;
+	// pub const QueueSize: usize = 2; // TODO: [AJ] this is apparently unused, can we remove?
```
Comment (thiolliere) [inline runtime/polkadot/src/lib.rs:836]: yes I think so https://github.com/paritytech/polkadot/pull/3553

File: xcm/src/v0/order.rs [hunk 5]
```diff
@@ -22,10 +22,11 @@ use derivative::Derivative;
 use parity_scale_codec::{self, Decode, Encode};
 
 /// An instruction to be executed on some or all of the assets in holding, used by asset-related XCM messages.
-#[derive(Derivative, Encode, Decode)]
+#[derive(Derivative, Encode, Decode, scale_info::TypeInfo)]
 #[derivative(Clone(bound = ""), Eq(bound = ""), PartialEq(bound = ""), Debug(bound = ""))]
 #[codec(encode_bound())]
 #[codec(decode_bound())]
+#[scale_info(skip_type_params(Call))]
```
Comment (thiolliere) [inline xcm/src/v0/order.rs:29]: it seems we can also do this no ? at least it is more consistent with codec bounds
```suggestion
#[scale_info(bounds(), skip_type_params(Call))]
```

File: bridges/bin/runtime-common/Cargo.toml (hunk 2 above)
Comment (thiolliere) [reply bridges/bin/runtime-common/Cargo.toml:14]: I added them to all I could find

**Review by thiolliere — APPROVED** [review]
* [x] maybe we can resolve this one https://github.com/paritytech/polkadot/pull/3336#discussion_r681009478

But otherwise it looks good to me

File: bridges/bin/runtime-common/Cargo.toml (hunk 2 above)
Comment (ascjones) [reply bridges/bin/runtime-common/Cargo.toml:14]: Thanks!

File: runtime/parachains/src/lib.rs (hunk 3 above)
Comment (ascjones) [reply runtime/parachains/src/lib.rs:25]: Removed this now and it seems to work okay now.

File: bridges/bin/millau/runtime/src/lib.rs (hunk 1 above)
Comment (HCastano) [reply bridges/bin/millau/runtime/src/lib.rs:463]: Yep, I can approve here in order to make the checks happy, but for a proper review of the bridges integration please open a PR in `parity-bridges-common`

**Comment by ascjones** [comment]
bot merge

**Review by ordian — APPROVED** [review]

**Comment by ordian** [comment]
bot merge

**Comment by ordian** [comment]
(fmt and spellcheck are ignored by bot)

**Comment by ordian** [comment]
the `check-transaction-versions` if failing though:
```
2021-09-15 12:50:29        API/INIT: Error: FATAL: Unable to initialize the API: Cannot convert metadata from version 14 to 13
    at EventEmitter.value (/usr/local/lib/node_modules/@polkadot/metadata-cmp/node_modules/@polkadot/api/base/Init.cjs:82:25)
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at runNextTicks (node:internal/process/task_queues:63:3)
    at processImmediate (node:internal/timers:437:9)
```

**Comment by jacogr** [comment]
Indeed. 2 issues -

1. The API PR has not been merged/released as of yet (so the comparator doesn't understand v14), see https://github.com/polkadot-js/api/pull/3759
2. The metadata compare has not been revamped for v14 (there is some work left since the type compares now need to be done differently), see https://github.com/polkadot-js/tools/issues/267

**Comment by bkchr** [comment]
@jacogr So, I assume we should disable the check until it works again?

**Comment by jacogr** [comment]
Hopefully sorted by Monday, aka need to get all released by then. But yes, would disable it at present.

**Comment by ascjones** [comment]
It seems the removed check is lingering.

**Comment by ordian** [comment]
@TriplEight ^^^

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**